python/ray/util/client/server/dataservicer.py

```diff
@@ -80,6 +84,44 @@ def fill_queue(
         output_queue.put(None)
 
 
+class ChunkCollector:
+    """
+    Helper class for collecting chunks from PutObject calls
+    """
+
+    def __init__(self):
+        self.curr_req_id = None
+        self.last_seen_chunk_id = -1
+        self.data = bytearray()
+
+    def add_chunk(self, req: ray_client_pb2.DataRequest):
+        if self.curr_req_id is not None and self.curr_req_id != req.req_id:
+            raise RuntimeError(
+                "Expected to receive a chunk from request with id "
+                f"{self.curr_req_id}, but found {req.req_id} instead."
+            )
+        self.curr_req_id = req.req_id
```

Comment:
Is this similar to what is used in #22100?

Reply:
Yup, depending on which one gets merged first there should be a way share this logic